File: webui/src/main/scala/com/taskadapter/webui/pages/WizardPanel.scala
package com.taskadapter.webui.pages

import com.taskadapter.vaadin14shim.Label
import com.taskadapter.webui.Page
import com.vaadin.flow.component.orderedlayout.{HorizontalLayout, VerticalLayout}
import com.vaadin.flow.component.progressbar.ProgressBar

import scala.collection.mutable

class WizardPanel {
  val layout = new VerticalLayout()
  layout.setMargin(true)
  val progressLabel = new Label()
  val progressBar = new ProgressBar()
  progressBar.setWidth("500px")
  val stepLayout = new HorizontalLayout()

  layout.add(progressLabel, progressBar)
  layout.add(stepLayout)

  val steps = mutable.Map[Int, WizardStep[_]]()

  def totalSteps = steps.size

  def showStep(step: Int): Unit = {
    progressLabel.setText(Page.message("newConfig.step", step + "", totalSteps + ""))
    progressBar.setValue(step.toFloat / totalSteps)
    stepLayout.removeAll()
    val previousStep = steps.get(step - 1)
    val config = previousStep.map(_.getResult).getOrElse("")
    val nextStep = steps(step)
    stepLayout.add(nextStep.ui(config))
  }

  def registerStep[C, RES](wizardStep: WizardStep[RES]): Unit = {
    val stepNumber = steps.size + 1
    steps += (stepNumber -> wizardStep)
  }

  def ui = layout

}

File: webui/src/main/java/com/taskadapter/webui/pages/config/wizard/WizardPanel.java
package com.taskadapter.webui.pages.config.wizard;

import com.taskadapter.webui.Page;
import com.taskadapter.webui.pages.WizardStep;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.progressbar.ProgressBar;

import java.util.HashMap;
import java.util.Map;

public class WizardPanel extends VerticalLayout {
    private final Map<Integer, WizardStep<?>> steps = new HashMap<>();

    private final Label progressLabel = new Label();
    private final ProgressBar progressBar = new ProgressBar();
    private final HorizontalLayout stepLayout = new HorizontalLayout();

    public WizardPanel() {
        setMargin(true);

        progressBar.setWidth("500px");

        add(progressLabel, progressBar,
                stepLayout);
    }

    private int totalSteps() {
        return steps.size();
    }

    public void showStep(int step) {
        progressLabel.setText(Page.message("newConfig.step", step + "", totalSteps() + ""));
        progressBar.setValue((float) step / totalSteps());
        stepLayout.removeAll();
        var previousStep = steps.get(step - 1);
        var config = previousStep == null ? null: previousStep.getResult();
        var nextStep = steps.get(step);
        stepLayout.add(nextStep.ui(config));
    }

    public void registerStep(WizardStep<?> wizardStep) {
        var stepNumber = steps.size() + 1;
        steps.put(stepNumber, wizardStep);
    }
}
